fix: change all doc links to permalinks

edna-harriet opened the pull request:
Closes issue #8171

**Motivation**
This PR replaces documentation and code references that previously pointed to the dev, main, master branches while also updating existing SHA commits with the release tags. However, references with SHA commits that would re-direct to broken page(Error 404) when replaced with tags, have not been changed.Those references still have their SHA commits.

**Description**
-Updated only .ts files containing github.com/ethereum/consensus-specs/blob/dev or main or master links.
-Also verified that no auto-generated file such as .js and .d.ts were modified.

**Steps to test or reproduce**
$ cd lodestar
$ git checkout issue-https://github.com/ChainSafe/lodestar/issues/8171
$ grep -Rnl "https://github.com/ethereum/consensus-specs/blob/[0-9a-f]\{7,40\}" --include='*.ts' --exclude='*.d.ts' --exclude='*.js' .

Conclusion
I invite review and I am ready to make request changes.

edna-harriet committed:
replaced branchname with latest SHA commit within github links in different files

edna-harriet committed:
modified github link with commits to use SHA commit

edna-harriet committed:
updated github commit links in files to use SHA commit

edna-harriet committed:
updated files with SHA commit

edna-harriet committed:
modified files with github links to use SHA commit

edna-harriet committed:
updated documents and code reference links to use SHA commit

nflaig commented on packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:
```diff
@@ -131,7 +131,8 @@ const MAX_ATTESTATIONS_PER_GROUP_ELECTRA = Math.min(
 );
 
-/** Same to https://github.com/ethereum/consensus-specs/blob/v1.5.0/specs/altair/beacon-chain.md#has_flag */
+//Reference : has_flag
```
why did you add this extra comment?

nflaig commented on packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:
```diff
@@ -131,7 +131,8 @@ const MAX_ATTESTATIONS_PER_GROUP_ELECTRA = Math.min(
 
-/** Same to https://github.com/ethereum/consensus-specs/blob/v1.5.0/specs/altair/beacon-chain.md#has_flag */
+//Reference : has_flag
+/** Same to https://github.com/ethereum/consensus-specs/blob/372e9e0ca53c066bee008f6fcb66ae1f94204ad0/specs/altair/beacon-chain.md#has_flag */
```
this spec reference was already using a tag before which is a permalink

nflaig commented on packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:
```diff
@@ -831,7 +832,8 @@ export function aggregateInto(attestation1: AttestationWithIndex, attestation2:
  * attestation data from different committee into a single attestation
- * https://github.com/ethereum/consensus-specs/blob/aba6345776aa876dad368cab27fbbb23fae20455/specs/_features/eip7549/validator.md?plain=1#L39
+ * Reference: Attestation
+ * https://github.com/ethereum/consensus-specs/blob/cf3c4eca8cff71de66153c1969ab66c1ea2c3338/specs/_features/eip7549/validator.md?plain=1#L39
```
I don't understand why this link is updated, it used a commit hash before

nflaig commented on packages/beacon-node/src/chain/opPools/utils.ts:
```diff
@@ -41,7 +41,8 @@ export function isValidBlsToExecutionChangeForBlockInclusion(
 ): boolean {
-  // For each condition from https://github.com/ethereum/consensus-specs/blob/dev/specs/capella/beacon-chain.md#new-process_bls_to_execution_change
+  // Reference : New process_bls_to_execution_change
+  // For each condition from https://github.com/ethereum/consensus-specs/blob/cf3eaab2b2c9039fc250db93bef7a3e92d49d594/specs/capella/beacon-chain.md#new-process_bls_to_execution_change
```
as suggested in https://github.com/ChainSafe/lodestar/issues/8171#issuecomment-3172560327 we wanna preferably use tags instead of commit hashes

matthewkeil commented:
Glad to see you getting this worked through.  Per our discussion on discord DM we would prefer to use spec release version tags.  I see that @nflaig has mentioned that above as well.  Please use this format
```
https://github.com/ethereum/consensus-specs/blob/v1.5.0/specs/altair/beacon-chain.md#has_flag
```
Also please address the Gemini comments and remove the added lines.  All that is necessary is the link itself.

matthewkeil commented on packages/beacon-node/src/chain/blocks/verifyBlocksExecutionPayloads.ts:
```diff
@@ -192,7 +192,8 @@ export async function verifyBlocksExecutionPayload(
     // However specs define this check to be run inside forkChoice's onBlock
-    // (https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/fork-choice.md#on_block)
+    // Reference : On_block
+    // (https://github.com/ethereum/consensus-specs/blob/46306c4f8addef64defc2a52d85f46992ba2421b/specs/bellatrix/fork-choice.md#on_block)
```
Can also remove the parentheses `(***)` that were there to clean things up further

edna-harriet replied:
> Can also remove the parentheses `(***)` that were there to clean things up further

@matthewkeil , most of the parenthesis initially in the files were used to comment the texts. So I just decided to leave them as they were.

edna-harriet committed:
reversed the spec reference to use the tag it had before and also removed the extra comment

edna-harriet committed:
changed the SHA commit on line 836, to previous SHA commit

edna-harriet commented:
@matthewkeil @nflaig , all your feedbacks are taken in. Am working on it. there are afew references that opens with SHA commit but when i change that to release tags(from the different available versions), the page breaks-Error 404. What do you advise I do in such case?

edna-harriet committed:
Replaced SHA commit references with their release tags

edna-harriet committed:
replaced references with SHA commit to use release tags

edna-harriet committed:
removed extra comments and replaced SHA commits with release tags

edna-harriet committed:
removed extra comments and changed SHA commits to release tags

nflaig commented:
> there are afew references that opens with SHA commit but when i change that to release tags(from the different available versions), the page breaks-Error 404. What do you advise I do in such case?

Links that are already using a commit sha are already permalinks, there is no need to update them

edna-harriet committed:
modified SHA commits to instead use release tags and removed extra comments

edna-harriet committed:
added release tags in place of SHA commit

edna-harriet committed:
replaced SHA commits with release tags

edna-harriet commented:
@nflaig @matthewkeil , I am ready for another review. I have implemented all AI and your requested changes.Thank you.

edna-harriet commented:
> > there are afew references that opens with SHA commit but when i change that to release tags(from the different available versions), the page breaks-Error 404. What do you advise I do in such case?
> 
> Links that are already using a commit sha are already permalinks, there is no need to update them

Noted

nflaig commented on packages/beacon-node/src/chain/bls/maybeBatch.ts:
```diff
@@ -39,7 +39,8 @@ export function verifySignatureSetsMaybeBatch(sets: SignatureSetDeserialized[]):
     // A signature could be malformed, in that case fromBytes throws error
     // blst-ts `verifyMultipleSignatures` is also a fallible operation if mul_n_aggregate fails
-    // see https://github.com/ChainSafe/blst-ts/blob/b1ba6333f664b08e5c50b2b0d18c4f079203962b/src/lib.ts#L291
+    // Reference : throw new ErrorBLST
```
why is this still added?

nflaig commented on packages/beacon-node/src/chain/bls/maybeBatch.ts:
```diff
@@ -39,7 +39,8 @@ export function verifySignatureSetsMaybeBatch(sets: SignatureSetDeserialized[]):
   } catch (_) {
     // A signature could be malformed, in that case fromBytes throws error
     // blst-ts `verifyMultipleSignatures` is also a fallible operation if mul_n_aggregate fails
-    // see https://github.com/ChainSafe/blst-ts/blob/b1ba6333f664b08e5c50b2b0d18c4f079203962b/src/lib.ts#L291
```
that's already a perma link and shouldn't be changed

nflaig commented on packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:
```diff
@@ -131,6 +131,7 @@ const MAX_ATTESTATIONS_PER_GROUP_ELECTRA = Math.min(
   MAX_ATTESTATIONS_ELECTRA
 );
 
+
```
this Pr shouldn't contain any other modifications, like adding extra new lines

matthewkeil replied:
All additions need to be removed

nflaig commented:
> @nflaig @matthewkeil , I am ready for another review. I have implemented all AI and your requested changes.Thank you.

it doesn't seem like it, there are still extra modifications and links that are already permalinks shouldn't be updated

edna-harriet committed:
removed extra comment

edna-harriet committed:
modified reference to the initial SHA commit value

edna-harriet committed:
removed extra comments

edna-harriet committed:
removed extra comments

edna-harriet commented:
@nflaig , i have effected the changes recommended above. Looking forward to reading your review .

edna-harriet commented:
@nflaig @matthewkeil , I worked on this and awaiting your review.

matthewkeil commented on packages/cli/src/cmds/validator/signers/importExternalKeystores.ts:
```diff
@@ -55,8 +55,6 @@ export function isVotingKeystore(filename: string): boolean {
     //
     // The format exported by the `staking-deposit-cli` library.
-    //
-    // Reference to function that generates keystores:
```
Why was this removed?  No changes should be included in this PR other than the links themselves...

matthewkeil commented:
> @nflaig @matthewkeil , I worked on this and awaiting your review.

I think you still have some review to do.  Please go through the `files` view of this PR to self audit where your work stands.
https://github.com/ChainSafe/lodestar/pull/8574/files

There are still a bunch of added empty lines, added words before the updated links (like `* Reference : on_payload`) and deleted existing comments.  The bar for inclusion into the codebase is high and needs a very detailed look.  Please make sure everything is 100% before marking this as ready for review.  Once you go through each of the comments above and resolve each, please click the resolve button on the comment.

edna-harriet commented:
> > @nflaig @matthewkeil , I worked on this and awaiting your review.
> 
> I think you still have some review to do. Please go through the `files` view of this PR to self audit where your work stands. https://github.com/ChainSafe/lodestar/pull/8574/files
> 
> There are still a bunch of added empty lines, added words before the updated links (like `* Reference : on_payload`) and deleted existing comments. The bar for inclusion into the codebase is high and needs a very detailed look. Please make sure everything is 100% before marking this as ready for review. Once you go through each of the comments above and resolve each, please click the resolve button on the comment.

@matthewkeil , i feel abit lost on this task. I know there are reference comments  that I added that should be removed eg the one you have cited above (Reference : on_payload). how about these ones where i replaced branch name with tag? what specifically am I supposed to correct about it?  Remove the //Mainnet config OR  leave the link with the dev branch as it was initially? Kindly shed some clarity on it?

// Mainnet config
// https://github.com/ethereum/consensus-specs/blob/dev/configs/mainnet.yaml
// https://github.com/ethereum/consensus-specs/blob/v1.6.0-alpha.6/configs/mainnet.yaml

matthewkeil commented on packages/beacon-node/src/chain/blocks/verifyBlocksExecutionPayloads.ts:
```diff
@@ -192,7 +192,7 @@ export async function verifyBlocksExecutionPayload(
     // a valid terminal PoW block.
     //
     // However specs define this check to be run inside forkChoice's onBlock
-    // (https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/fork-choice.md#on_block)
```
This is what a correct change should look like.  For the most part, all other diffs are incorrect.

matthewkeil commented on packages/beacon-node/src/chain/validation/attestation.ts:
```diff
@@ -764,7 +765,7 @@ function verifyAttestationTargetRoot(headBlock: ProtoBlock, targetRoot: Root, at
     //
-    // Reference:
+  
     // https://github.com/ethereum/consensus-specs/pull/2001#issuecomment-699246659
```
This comment was existing and should be put back

matthewkeil commented:
> > > @nflaig @matthewkeil , I worked on this and awaiting your review.
> > 
> > 
> > I think you still have some review to do. Please go through the `files` view of this PR to self audit where your work stands. https://github.com/ChainSafe/lodestar/pull/8574/files
> > There are still a bunch of added empty lines, added words before the updated links (like `* Reference : on_payload`) and deleted existing comments. The bar for inclusion into the codebase is high and needs a very detailed look. Please make sure everything is 100% before marking this as ready for review. Once you go through each of the comments above and resolve each, please click the resolve button on the comment.
> 
> @matthewkeil , i feel abit lost on this task. I know there are reference comments that I added that should be removed eg the one you have cited above (Reference : on_payload). how about these ones where i replaced branch name with tag? what specifically am I supposed to correct about it? Remove the //Mainnet config OR leave the link with the dev branch as it was initially? Kindly shed some clarity on it?
> 
> // Mainnet config // https://github.com/ethereum/consensus-specs/blob/dev/configs/mainnet.yaml // https://github.com/ethereum/consensus-specs/blob/v1.6.0-alpha.6/configs/mainnet.yaml

I added a couple notes above. I hope that helps.  Please also make sure to click "resolve" for the gemini created issues if they are fixed.

edna-harriet commented:
@matthewkeil, I have read the notes but I feel confused now that you have said more diffs are incorrect. I followed your earlier suggested convention here :  https://github.com/ethereum/consensus-specs/blob/v1.5.0/specs/altair/beacon-chain.md#has_flag.

By tags for permalink, did you mean : v1.5.0  or   #has_flag . I focused on using for example,  like v1.5.0 and seemingly I did not get it right.

Would you mind scheduling a call so that you may explain to me? I feel it's some concept about addressing this issue that I have not yet grasped. 


On Thu, Oct 30, 2025 at 7:47 PM Matthew Keil ***@***.***>
wrote:

> *matthewkeil* left a comment (ChainSafe/lodestar#8574)
> <https://github.com/ChainSafe/lodestar/pull/8574#issuecomment-3469002462>
>
> @nflaig <https://github.com/nflaig> @matthewkeil
> <https://github.com/matthewkeil> , I worked on this and awaiting your
> review.
>
> I think you still have some review to do. Please go through the files
> view of this PR to self audit where your work stands.
> https://github.com/ChainSafe/lodestar/pull/8574/files
> There are still a bunch of added empty lines, added words before the
> updated links (like * Reference : on_payload) and deleted existing
> comments. The bar for inclusion into the codebase is high and needs a very
> detailed look. Please make sure everything is 100% before marking this as
> ready for review. Once you go through each of the comments above and
> resolve each, please click the resolve button on the comment.
>
> @matthewkeil <https://github.com/matthewkeil> , i feel abit lost on this
> task. I know there are reference comments that I added that should be
> removed eg the one you have cited above (Reference : on_payload). how about
> these ones where i replaced branch name with tag? what specifically am I
> supposed to correct about it? Remove the //Mainnet config OR leave the link
> with the dev branch as it was initially? Kindly shed some clarity on it?
>
> // Mainnet config //
> https://github.com/ethereum/consensus-specs/blob/dev/configs/mainnet.yaml
> //
> https://github.com/ethereum/consensus-specs/blob/v1.6.0-alpha.6/configs/mainnet.yaml
>
> I added a couple notes above. I hope that helps
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ChainSafe/lodestar/pull/8574#issuecomment-3469002462>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGJW7KHGBWKFFGIZQTD4XL32I6K7AVCNFSM6AAAAACKJUM76KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTINRZGAYDENBWGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nflaig commented:
@edna-harriet all you have to do for this issue is search the code for `consensus-specs/blob/dev/specs` and replace those with a tag instead of referencing `dev` branch

any other modifications like extra new lines, reformatting need to be avoided

edna-harriet committed:
removed extra lines and returned comments removed erroneously

edna-harriet committed:
removed extra lines, comments and returned initial SHA commit that had been replaced with a tag

edna-harriet committed:
removed extra lines and comments

edna-harriet committed:
modified extra lines and changes previously made

edna-harriet committed:
removed extra comments

edna-harriet committed:
moved suggested tags back to initial SHA commits or branchnames

edna-harriet committed:
removed added ln77

edna-harriet committed:
undo:removed added ln77

edna-harriet committed:
replaced branchnames with tags

edna-harriet committed:
used tags in place of branchnames

edna-harriet committed:
replaced branch names with tags

edna-harriet committed:
use tags in place of branch names

edna-harriet committed:
replaced branch names with tag

edna-harriet committed:
returned comment that had been remove

edna-harriet committed:
reversed SHA commit back to the initial SHA commit value

nflaig commented:
I don't understand @edna-harriet why you marked this as ready for review, please look at the diff yourself

<img width="137" height="50" alt="image" src="https://github.com/user-attachments/assets/e0513ee1-9301-452a-b99f-73d36e6cdf14" />


there are still random and wrong modifications

<img width="719" height="274" alt="image" src="https://github.com/user-attachments/assets/6311953a-ac7d-4222-8aba-99b16b1ce5b5" />


and the branch has merge conflicts

<img width="568" height="129" alt="image" src="https://github.com/user-attachments/assets/d4d2ceca-4a77-4d79-a28b-0e109f4d5d38" />


I am going ahead and close this because we don't have time to do that many review cycles and issues are not being addressed and new issues come up each time you push changes
